packages/firestore/test/unit/local/query_engine.test.ts

```diff
@@ -403,7 +485,85 @@ describe('QueryEngine', () => {
       doc('coll/b', 1, { order: 3 })
     ]);
   });
-});
+
+  if (!durable) {
+    return;
+  }
+
+  if (!INDEXING_ENABLED) {
+    return;
+  }
+
+  it('combines indexed with non-indexed results', async () => {
+    debugAssert(durable, 'Test requires durable persistence');
+
+    const doc1 = doc('coll/a', 1, { 'foo': true });
+    const doc2 = doc('coll/b', 2, { 'foo': true });
+    const doc3 = doc('coll/c', 3, { 'foo': true });
+    const doc4 = doc('coll/d', 3, { 'foo': true }).setHasLocalMutations();
+
+    await indexManager.addFieldIndex(
+      fieldIndex('coll', { fields: [['foo', IndexKind.ASCENDING]] })
+    );
+
+    await addDocument(doc1);
+    await addDocument(doc2);
+    await indexManager.updateIndexEntries(documentMap(doc1, doc2));
+    await indexManager.updateCollectionGroup(
+      'coll',
+      newIndexOffsetFromDocument(doc2)
+    );
+
+    await addDocument(doc3);
+    await addMutation(setMutation('coll/d', { 'foo': true }));
+
+    const queryWithFilter = queryWithAddedFilter(
+      query('coll'),
+      filter('foo', '==', true)
+    );
+    const results = await expectOptimizedCollectionQuery(() =>
+      runQuery(queryWithFilter, SnapshotVersion.min())
+    );
+
+    verifyResult(results, [doc1, doc2, doc3, doc4]);
+  });
+
+  it('uses partial index for limit queries', async () => {
+    debugAssert(durable, 'Test requires durable persistence');
+
+    const doc1 = doc('coll/1', 1, { 'a': 1, 'b': 0 });
+    const doc2 = doc('coll/2', 1, { 'a': 1, 'b': 1 });
+    const doc3 = doc('coll/3', 1, { 'a': 1, 'b': 2 });
+    const doc4 = doc('coll/4', 1, { 'a': 1, 'b': 3 });
+    const doc5 = doc('coll/5', 1, { 'a': 2, 'b': 3 });
+    await addDocument(doc1, doc2, doc3, doc4, doc5);
+
+    await indexManager.addFieldIndex(
+      fieldIndex('coll', { fields: [['a', IndexKind.ASCENDING]] })
+    );
+    await indexManager.updateIndexEntries(
+      documentMap(doc1, doc2, doc3, doc4, doc5)
+    );
+    await indexManager.updateCollectionGroup(
+      'coll',
+      newIndexOffsetFromDocument(doc5)
+    );
+
+    const q = queryWithLimit(
+      queryWithAddedFilter(
+        queryWithAddedFilter(query('coll'), filter('a', '==', 1)),
+        filter('b', '==', 1)
+      ),
+      3,
+      LimitType.First
+    );
+    const results = await expectOptimizedCollectionQuery(() =>
+      runQuery(q, SnapshotVersion.min())
+    );
+
+    verifyResult(results, [doc2]);
+  });
+}
```

Comment:
We should also add a test for refilling, unless there is already one. It's late for me so I did not bother to check.